apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -295,590 +418,952 @@ poll(Node, ReturnAddr, Shard, Iterator, Opts = #{timeout := Timeout}) ->
-    },
-    Req2 = #poll_req{
-        return_addr = Raddr,
-        msg_matcher = fun(_, _) -> false end,
-        it = {it2, <<"1">>}
-    },
-    Req3 = #poll_req{
-        return_addr = Raddr,
-        msg_matcher = fun(_, _) -> true end,
-        it = {it3, <<"2">>}
-    },
-    %% Messages:
-    M1 = {<<"1">>, #message{id = <<"1">>}},
-    M2 = {NextKey, #message{id = <<"2">>}},
-    Reqs = [Req1, Req2, Req3],
-    [
-        ?_assertMatch(
-            #beam{
-                iterators = [],
-                pack = [
-                    {<<"1">>, <<>>, #message{id = <<"1">>}},
-                    {_Next, <<>>, #message{id = <<"2">>}}
-                ]
-            },
-            pack(UpdateIterator, NextKey, [], [M1, M2])
-        ),
-        ?_assertMatch(
-            #beam{
-                iterators = [
-                    {Raddr, {it1, NextKey}}, {Raddr, {it2, NextKey}}, {Raddr, {it3, NextKey}}
-                ],
-                pack = [
-                    {<<"1">>, <<1:1, 0:1, 1:1>>, #message{id = <<"1">>}},
-                    {NextKey, <<1:1, 0:1, 1:1>>, #message{id = <<"2">>}}
-                ]
-            },
-            pack(UpdateIterator, NextKey, Reqs, [M1, M2])
-        )
-    ].
-
-split_test_() ->
-    M1 = {<<"1">>, #message{id = <<"1">>}},
-    M2 = {<<"2">>, #message{id = <<"2">>}},
-    M3 = {<<"3">>, #message{id = <<"3">>}},
-    Its = [{<<"it1">>, it1}, {<<"it2">>, it2}],
-    Beam1 = #beam{
-        iterators = Its,
-        pack = [
-            {<<"1">>, <<1:1, 0:1>>, element(2, M1)},
-            {<<"2">>, <<0:1, 1:1>>, element(2, M2)},
-            {<<"3">>, <<1:1, 1:1>>, element(2, M3)}
-        ]
-    },
-    [{<<"it1">>, Result1}, {<<"it2">>, Result2}] = lists:sort(split(Beam1)),
-    [
-        ?_assertMatch(
-            {ok, it1, [M1, M3]},
-            Result1
-        ),
-        ?_assertMatch(
-            {ok, it2, [M2, M3]},
-            Result2
-        )
-    ].
+-compile({inline, gvar/1}).
+-spec gvar(dbshard()) -> gvar().
+gvar(DBShard) ->
+    persistent_term:get(?pt_gvar(DBShard)).
+
+%% subtab(Shard) ->
+%%     persistent_term:get(?subtid(Shard)).
+
+make_fctab() ->
+    ets:new(emqx_ds_beamformer_flow_control_tab, [
+        set,
+        protected,
+        {keypos, #fctab.sub_ref},
+        {read_concurrency, true}
+    ]).
+
+make_owner_tab() ->
+    ets:new(emqx_ds_beamformer_owner_tab, [
+        set,
+        public,
+        {keypos, #owner_tab.sub_ref},
+        {write_concurrency, true}
+    ]).
+
+%% @doc Send fake monitor messages to all subscribers registered in
+%% the table, nudging them to resubscribe.
+on_worker_down(InheritedSubTab) ->
+    MS = {#sub_state{client = '$1', req_id = '$2', _ = '_'}, [], [{{'$1', '$2'}}]},
+    notify_down_loop(ets:select(InheritedSubTab, [MS], 100)).
+
+notify_down_loop('$end_of_table') ->
+    ok;
+notify_down_loop({Matches, Cont}) ->
+    _ = [Pid ! {'DOWN', Ref, process, self(), worker_crash} || {Pid, Ref} <- Matches],
```

Comment:
Nit: if we're faking a `DOWN` signal from the worker that has just died (and beamformer received in the real `DOWN` signal), wouldn't it be more "realistic" to use the worker pid here instead of beamformer's?

Reply:
The intention is to hide the presence of workers from the client.